### PR TITLE
Preparing for adding kubevirt samples

### LINKED ISSUE #43: Support for automated testing for k8s and odr
Currently we have one application (busybox) and kustomizations to deploy it manually on regional-dr or metro-dr using rbd or cephfs storage.

For automated testing we need subscription kustomization for each application variant (rdr-rbd, rdr-cephfs, mdr-rbd).

The layout should make it easy to add more applications (e.g. busybox statefulset, busybox daemonset, kubevirt vms with pvc, data-volume, or data-volume-template).

## How it should work

- Nothing is shared between applications, so we can deploy and undeploy multiple applications concurrently. Example use case - test all applications variants concurrently on the same cluster.
- Split application resources and ocm resources. Application resources are different for each application. All application can share the same subscription by kustomizing the subscription path and channel.
- Add kustomization for particular DR type (rdr, mdr), storage type (rbd, cephfs) and cluster type (odr, k8s).
- Uniform names so we can build applications and subscription names programatically using a template
  (e.g. "subscription/{platform}/{application}-{topology}-{storage}")
- Don't include the drpc - current ramen tests generate the drpc since it needs to many changes (drpolicy, namespace, subscription, placement, pvc selector)

## Suggested layout

```
busybox-deployment/
    odr-rdr-rbd/
        kustomization.yaml
    odr-rdr-cephfs/
        kustomization.yaml
    odr-mdr-rbd/
        kustomization.yaml
    k8s-rdr-rbd/
        kustomization.yaml
    deployment.yaml
    kustomization.yaml
    pvc.yaml
subscription/
    odr/
        busybox-deployment-rdr-rbd/
            kustomization.yaml
        busybox-deployment-rdr-cephfs/
            kustomization.yaml
        busybox-deployment-mdr-rbd/
            kustomization.yaml
    k8s/
        busybox-deployment-rdr-rbd/
            kustomization.yaml
    binding.yaml
    channel.yaml
    kustomization.yaml
    namespace.yaml
    placement.yaml
    subscription.yaml
```

## Usage in OpenShift console

When selecting application **path** use one of:

```
busybox-deployment/odr-rdr-rbd
busybox-deployment/odr-rdr-cephfs
busybox-deployment/odr-mdr-rbd
```
## Usage in automated tests

When deploying a subscription in automated tests use one of these:

Subscriptions for odr tests:

```
subscription/odr/busybox-deployment-rdr-rbd
subscription/odr/busybox-deployment-rdr-cephfs
subscription/odr/busybox-deployment-mdr-rbd
```

Subscriptions for k8s tests:

```
subscription/k8s/busybox-deployment-rdr-rbd
```

### PR DESCRIPTION
**Pull request by nirs**
Preparing for adding kubevirt samples, clean up the current samples, and change the directory layout to make it easy to add new applications and allow automated testing on all cluster and storage types.

Status:
- [x] Tested using basic-test with https://github.com/RamenDR/ramen/pull/1133
- [x] Manual deployment on ocp

Fixes #43

### CODE REVIEW

File: dr/core/base/drpc.yaml [hunk 1]
```diff
@@ -2,16 +2,14 @@
...
+  preferredCluster: cluster1
   drPolicyRef:
     name: dr-policy
   placementRef:
     kind: Placement
-    name: busybox-placement
+    name: placement
   pvcSelector:
     matchLabels:
       appname: busybox
```
Comment (ShyamsundarR) [inline dr/core/base/drpc.yaml:15]: Label needs to be kustoimized in `dr/core/deployment-k8s-regional-rbd/kustomization.yaml` to `app: deployment-rbd`?

File: channel/kustomization.yaml [hunk 2]
```diff
@@ -1,3 +1,4 @@
+---
 resources:
-- channel.yaml
-- namespace.yaml
+  - channel.yaml
```
Comment (ShyamsundarR) [inline channel/kustomization.yaml:3]: I am not sure we need this indent for lists, from what I can [see](https://yaml.org/spec/1.2.2/#21-collections) we should actually avoid it. Is there a reason this is indented here and in other files where lists are in use?

**Review by ShyamsundarR — COMMENTED** [review]
A few questions:
- Why is there a `core` subdirectory within `dr` and `subscription` directories?
- My understanding of `core` under root, and `deployment` under the same is to have another higher level directory like `kubevirt` with its variants within. This reduces some clutter (so that we avoid deployment/sts/daemonset within core). The issue I see is with the name `core` or just stashing away 3 workload types from the clutter (we will still need kubevirt at the root with it's sub-directories in there). It would be preferable to be more explicit and have deployment at the root. We can add a `workloads` or similar directory at the root and move all workloads within it (avoid clutter at root of the repo).

The second is more key than the first, as it impacts product documentation.

**Comment by nirs** [comment]
> A few questions:
> 
> * Why is there a `core` subdirectory within `dr` and `subscription` directories?

We need different drpc for busybox and kubevirt, mainly because we use different pvc selector, based on appname label. We use the same drpc if we use the same appname for all samples, maybe `appname: sample`

For subscription I assumed we will have few differences between core and kubvrit, but I managed to share the same files in #45.

If we remove core and use the layout bellow, we have something like:

```
subscripiton/deployment-...
subscription/kubevirt-vm-pvc-...
```

And same for dr/

> * My understanding of `core` under root, and `deployment` under the same is to have another higher level directory like `kubevirt` with its variants within. This reduces some clutter (so that we avoid deployment/sts/daemonset within core). The issue I see is with the name `core` or just stashing away 3 workload types from the clutter (we will still need kubevirt at the root with it's sub-directories in there). It would be preferable to be more explicit and have deployment at the root. We can add a `workloads` or similar directory at the root and move all workloads within it (avoid clutter at root of the repo).
> 
> The second is more key than the first, as it impacts product documentation.

We already discussed this structure last week - grouping the standard resources (deployment, statefulset, daemonset under core, and 3 vm samples under kubevirt. For kubvirt grouping the VMs makes sense since we want to have a kubvirt specific README.md. For core samples, I don't think we need any specific README.md.

Do you prefer this layout?

```
workloads/
  deployment/
  statefulset/
  daemonset/
  kubevirt/
  kafka/
  mongodb/
```

From OpenShift console, adding user will need to choose paths like:

- workloads/deployemnt/odr-regional-rbd
- workloads/kubevirt/vm-pvc/odr-regional-rbd

File: dr/core/base/drpc.yaml (hunk 1 above)
Comment (nirs) [reply dr/core/base/drpc.yaml:15]: Yes, the label is customized there.

File: channel/kustomization.yaml (hunk 2 above)
Comment (nirs) [reply channel/kustomization.yaml:3]: We need this to keep yamllint happy.

File: README.md [hunk 3]
```diff
@@ -5,89 +5,165 @@ OCM Stateful application samples, including Ramen resources.
...
+
+   This creates a DRPlacementControl resource for the busybox deployment
+   in the `deployment-rbd` namespace and can be viewed using:
+
+   ```
+   kubectl get drpc -n deployment-rbd
+   ```
 
-1. Ensure Placement is pointing to the cluster where the workload is
+   At this point the application is managed by *Ramen*.
```
Comment (ShyamsundarR) [inline README.md:141]: The old statement can be better, this should be something like, `At this point the placement of the application is managed by *Ramen*`

File: dr/core/base/drpc.yaml (hunk 1 above)
Comment (ShyamsundarR) [reply dr/core/base/drpc.yaml:15]: I am missing something here, the spec.pvcSelector should be customized in the file pointed to above, based on the commonLabels that is being set? I would assume the common labels is carried over to the actual workload artifacts as well (for e.g the `app` label). If not then this cannot (yet?) be customized here.

File: channel/kustomization.yaml (hunk 2 above)
Comment (ShyamsundarR) [reply channel/kustomization.yaml:3]: k8s outputs sequences with indentation of the parent node, controller-sdk does the same etc. I am "thinking" we modify the yamllint to add the [rule](https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.indentation) for `indent-sequences`.

File: dr/core/base/drpc.yaml (hunk 1 above)
Comment (nirs) [reply dr/core/base/drpc.yaml:15]: pvcSelector is not customized - we customize appname in the kustomization.yaml to match the hardcoded value in the drpc.

File: channel/kustomization.yaml (hunk 2 above)
Comment (nirs) [reply channel/kustomization.yaml:3]: Default yamllint rules creates more readable yaml.

**Comment by nirs** [comment]
Conclusions after discussion:
- Go with `workloads{deployment,kubevirt}` layout, removing core
- Otherwise current layout is fine

File: README.md (hunk 3 above)
Comment (nirs) [reply README.md:141]: Fixed in next version.

**Pull request merged**